File: Leetcode/617.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs(root1 *TreeNode, root2 *TreeNode) {
    (*root1).Val += (*root2).Val
    if (*root1).Left != nil && (*root2).Left != nil {
        dfs((*root1).Left, (*root2).Left)
    } else if (*root1).Left == nil && (*root2).Left != nil {
        (*root1).Left = (*root2).Left
    }
    if (*root1).Right != nil && (*root2).Right != nil {
        dfs((*root1).Right, (*root2).Right)
    } else if (*root1).Right == nil && (*root2).Right != nil {
        (*root1).Right = (*root2).Right
    }
}

func mergeTrees(root1 *TreeNode, root2 *TreeNode) *TreeNode {
    if root1 == nil && root2 == nil {
        return nil
    } else if root1 != nil && root2 == nil {
        return root1
    } else if root1 == nil && root2 != nil {
        return root2
    }
    dfs(root1, root2)
    return root1
}

File: Leetcode/617/617.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // merge root2 to root1
    void dfs(TreeNode* root1, TreeNode *root2) {
        root1->val += root2->val;
        if (root1->left && root2->left) {
            dfs(root1->left, root2->left);
        }
        else if (!root1->left && root2->left) {
            root1->left = root2->left;
        }
        if (root1->right && root2->right) {
            dfs(root1->right, root2->right);
        }
        else if (!root1->right && root2->right) {
            root1->right = root2->right;
        }
    }
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        if (!root1 && !root2) return nullptr;
        else if (root1 && !root2) return root1;
        else if (!root1 && root2) return root2;
        dfs(root1, root2);
        return root1;
    }
};